File: Arrays-And-Strings/Challenge-2.js
//O(N log N) time complexity and O(N) space complexity

const permutationChecker = (string1, string2) => {
  if (string1.length !== string2.length) {
    return false;
  }
  const sortedStringOne = string1.split('').sort().join('');
  const sortedStringTwo = string2.split('').sort().join('');

  return sortedStringOne === sortedStringTwo ? true : false;
};

console.log(permutationChecker('abd', 'dba')); //true
console.log(permutationChecker('abcsdaease', 'ffe')); //false
console.log(permutationChecker('', '')); // true

//Using a hashmap

//O(N) time complexity and O(N) space complexity

const permutationChecker2 = (string1, string2) => {
  if (string1.length !== string2.length) {
    return false;
  }
  const hashMap = {};
  for (let i = 0; i < string2.length; i++) {
    const character = string2[i];
    if (hashMap[character]) {
      hashMap[character] += 1;
    } else {
      hashMap[character] = 1;
    }
  }
  for (let i = 0; i < string1.length; i++) {
    const character = string1[i];
    if (!hashMap[character]) {
      return false;
    } else if (hashMap[character] > 0) {
      hashMap[character] -= 1;
    } else {
      return false;
    }
  }
  return true;
};

console.log(permutationChecker2('', '')); // true
console.log(permutationChecker2('hashh', 'sahhh')); // true
console.log(permutationChecker2('a', 'c')); // false
console.log(permutationChecker2('add', 'abd')); // false
console.log(permutationChecker2('adsadas', 'xx')); // false

File: Arrays-And-Strings/Challenge-2.py
# O(N log N) time complexity and O(N) space complexity

def permutation_checker(string1, string2):
  if len(string1) is not len(string2):
    return False

  sorted_string_one = "".join(sorted(string1))
  sorted_string_two = "".join(sorted(string2))

  return sorted_string_one == sorted_string_two

print(permutation_checker('sbd','dbbs')) # False
print(permutation_checker('abc','cba')) # True

def permutation_checker2(string1,string2):
  if len(string1) is not len(string2):
    return False

  dictionary = {}

  for character in string2:
    if character in dictionary:
      dictionary[character] += 1
    else:
      dictionary[character] = 1

  for character in string1:
    if character in dictionary:
      dictionary[character] -= 1
    else:
      return False
  return True

print(permutation_checker2('sbd','dbbs')) # False
print(permutation_checker2('abc','cba')) # True

